cmd/segment-reaper/observer_test.go

```diff
@@ -0,0 +1,613 @@
+					},
+				},
+			}
+		}
+
+		var (
+			buf   = &bytes.Buffer{}
+			obsvr = observer{
+				db:      teststore.New(),
+				writer:  csv.NewWriter(buf),
+				objects: bucketsObjs,
+			}
+		)
+
+		ctx := testcontext.New(t)
+		defer ctx.Cleanup()
+
+		err := obsvr.analyzeProject(ctx.Context)
+		require.NoError(t, err)
+
+		// TODO: Add assertions for buf content
+	})
+
+	t.Run("object with unencrypted segments with different stored number", func(t *testing.T) {
+		var bucketsObjs bucketsObjects
+		{ // Generate an object
+			const (
+				bucketName = "analyzeBucket"
+				objPath    = storj.Path("analyzePath")
+			)
+			var (
+				segments           uint64
+				invalidNumSegments byte
+			)
+			{
+				numSegments := rand.Intn(62) + 1
+				segments = math.MaxUint64 >> numSegments
+
+				for {
+					numSeg := rand.Intn(65)
+					if numSeg != numSegments {
+						invalidNumSegments = byte(numSeg)
+						break
+					}
+				}
+			}
+
+			bucketsObjs = bucketsObjects{
+				bucketName: map[storj.Path]*object{
+					objPath: {
+						segments:                 segments,
+						expectedNumberOfSegments: invalidNumSegments,
+						hasLastSegment:           false,
+					},
+				},
+			}
+		}
+
+		var (
+			buf   = &bytes.Buffer{}
+			obsvr = observer{
+				db:      teststore.New(),
+				writer:  csv.NewWriter(buf),
+				objects: bucketsObjs,
+			}
+		)
+
+		ctx := testcontext.New(t)
+		defer ctx.Cleanup()
+
+		err := obsvr.analyzeProject(ctx.Context)
+		require.NoError(t, err)
+
+		// TODO: Add assertions for buf content
+	})
+}
+
+// objectSegmentRef is an object segment reference to be used for simulating
+// calls observer.processSegment
+type objectSegmentRef struct {
+	path    metainfo.ScopedPath
+	pointer *pb.Pointer
+}
+
+// crateNewObjectSegments creates a list of segment references which belongs to
+// a same object.
+//
+// If inline is true the last segment will be of INLINE type.
+//
+// If withNumSegments is true the last segment pointer will have the
+// NumberOfSegments set.
+//
+// It returns the object path and the list of object segment references.
+func createNewObjectSegments(t *testing.T, numSegments int, projectID *uuid.UUID, bucketName string, inline bool, withNumSegments bool) (objectPath string, _ []objectSegmentRef) {
+	t.Helper()
+
+	var objectID string
+	{
+		id := testrand.UUID()
+		objectID = id.String()
+	}
+
+	var (
+		projectIDString = projectID.String()
+		references      = make([]objectSegmentRef, 0, numSegments)
+		encryptedPath   = fmt.Sprintf("%s-%s-%s", projectIDString, bucketName, objectID)
+	)
+
+	for i := 0; i < (numSegments - 1); i++ {
+		references = append(references, objectSegmentRef{
+			path: metainfo.ScopedPath{
+				ProjectID:           *projectID,
+				ProjectIDString:     projectIDString,
+				BucketName:          bucketName,
+				Segment:             fmt.Sprintf("s%d", i),
+				EncryptedObjectPath: encryptedPath,
+				Raw:                 fmt.Sprintf("%s/%s/%s/s%d", projectIDString, bucketName, objectID, i),
```

Comment:
Is this the correct raw path structure? I thought the segment number is the second component in the path, not the last one, i.e. `<project-id>/<segment-number>/<bucket>/<object-path>`.

Reply:
Probably you're right, I mostly created a raw path which is unique per segment and object.

The current `processSegment` doesn't care about the raw path value, so it's fine for the test.

If you confirm that's the correct form, I will change it for not causing confusion.

Reply:
`<project-id>/<segment-number>/<bucket>/<object-path>` is a valid path structure.

Reply:
We even have a handy function to create the segment path: https://github.com/storj/storj/blob/7288d0178138338a72bb732e520c3f2f5bbc92c8/satellite/metainfo/metainfo.go#L641

Reply:
yup, I gonna use it

Reply:
Done